# Single-Qubit Gate Sequence and Statevector Analysis

## 1. Objective
The objective of this experiment is to study the cumulative effect of applying multiple single-qubit quantum gates in series to a single qubit initially prepared in the computational ground state $|0\rangle$, and to examine the resulting quantum statevector without performing measurement collapse.

---

## 2. Quantum Gates Used
This experiment applies a cascade of five distinct single-qubit unitary operators:

1. **H (Hadamard Gate)**:
   Creates an equal superposition from computational-basis states:
   $$H = \frac{1}{\sqrt{2}}\begin{pmatrix} 1 & 1 \\ 1 & -1 \end{pmatrix}, \quad H|0\rangle = \frac{|0\rangle + |1\rangle}{\sqrt{2}} = |+\rangle$$

2. **S (Phase Gate / $\sqrt{Z}$)**:
   Applies a relative phase shift of $\pi/2$ (multiplying by $e^{i\pi/2} = i$) to the $|1\rangle$ component:
   $$S = \begin{pmatrix} 1 & 0 \\ 0 & i \end{pmatrix}, \quad S|1\rangle = i|1\rangle$$

3. **Y (Pauli-Y Gate)**:
   Performs a combined bit-and-phase transformation:
   $$Y = \begin{pmatrix} 0 & -i \\ i & 0 \end{pmatrix}, \quad Y|0\rangle = i|1\rangle, \quad Y|1\rangle = -i|0\rangle$$

4. **T (T Gate / $\pi/8$ Gate / $\sqrt{S}$)**:
   Applies a relative phase shift of $\pi/4$ (multiplying by $e^{i\pi/4} = \frac{1+i}{\sqrt{2}}$) to the $|1\rangle$ component:
   $$T = \begin{pmatrix} 1 & 0 \\ 0 & e^{i\pi/4} \end{pmatrix}, \quad T|1\rangle = e^{i\pi/4}|1\rangle$$

5. **Z (Pauli-Z Gate / Phase-Flip)**:
   Applies a relative phase flip of $\pi$ (multiplying by $-1 = e^{i\pi}$) to the $|1\rangle$ component while leaving $|0\rangle$ unchanged:
   $$Z = \begin{pmatrix} 1 & 0 \\ 0 & -1 \end{pmatrix}, \quad Z|0\rangle = |0\rangle, \quad Z|1\rangle = -|1\rangle$$

---

## 3. Gate Sequence and Transformation Pipeline
The gates are applied sequentially to the same qubit:

$$|0\rangle \xrightarrow{\quad H \quad} |+\rangle \xrightarrow{\quad S \quad} |+i\rangle \xrightarrow{\quad Y \quad} |+i\rangle \xrightarrow{\quad T \quad} |\psi_T\rangle \xrightarrow{\quad Z \quad} |\psi_{\text{final}}\rangle$$

Because quantum operations are unitary transformations, each gate acts directly on the state produced by the preceding operation:
$$|\psi_{\text{final}}\rangle = Z \cdot T \cdot Y \cdot S \cdot H |0\rangle$$

### Environment Setup
Install Qiskit if executing in a fresh environment.

In [1]:
import sys
!{sys.executable} -m pip install qiskit


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 9.7/9.7 MB 47.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.4/2.4 MB 59.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 54.5/54.5 kB 2.3 MB/s eta 0:00:00


## 4. Implementation
- **Circuit Creation**: `QuantumCircuit(1)` creates a one-qubit quantum circuit initialized in the ground state $|0
angle$.
- **Sequence Definition**: The `gate_sequence` list stores tuples containing each gate name and its corresponding Qiskit method:
  `gate_sequence = [("H", qc.h), ("S", qc.s), ("Y", qc.y), ("T", qc.t), ("Z", qc.z)]`
- **Iterative Execution**: A loop applies each gate sequentially to qubit 0 while printing the gate label.
- **Analytical Statevector Calculation**: `Statevector.from_instruction(qc)` calculates the exact mathematical statevector resulting from the full sequence without performing projective measurements.

In [2]:
from qiskit import QuantumCircuit
from qiskit.quantum_info import Statevector

qc = QuantumCircuit(1)

gate_sequence = [
    ("H", qc.h),
    ("S", qc.s),
    ("Y", qc.y),
    ("T", qc.t),
    ("Z", qc.z)
]

print("Applied Gates:")

for name, gate in gate_sequence:
    print(name, end=" ")
    gate(0)

state = Statevector.from_instruction(qc)

print("\n\nCircuit")
print(qc.draw())

print("\nStatevector")
print(state)


Applied Gates:
H S Y T Z 

Circuit
   ┌───┐┌───┐┌───┐┌───┐┌───┐
q: ┤ H ├┤ S ├┤ Y ├┤ T ├┤ Z ├
   └───┘└───┘└───┘└───┘└───┘

Statevector
Statevector([0.70710678+0.j , 0.5       -0.5j],
            dims=(2,))


## 5. Results & Observations

### Circuit Structure
The generated circuit diagram displays the sequential application of the five single-qubit gates on qubit `q`:
```
   ┌───┐┌───┐┌───┐┌───┐┌───┐
q: ┤ H ├┤ S ├┤ Y ├┤ T ├┤ Z ├
   └───┘└───┘└───┘└───┘└───┘
```

### Final Statevector
The statevector produced by the execution is:
$$\text{Statevector}([0.70710678+0.\text{j}, \quad 0.5 - 0.5\text{j}], \quad \text{dims}=(2,))$$

- **Cumulative Step-by-Step Evolution**:
  1. $H|0\rangle = \frac{1}{\sqrt{2}}(|0\rangle + |1\rangle) = |+\rangle$.
  2. $S|+\rangle = \frac{1}{\sqrt{2}}(|0\rangle + i|1\rangle) = |+i\rangle$.
  3. $Y|+i\rangle = |+i\rangle$ (since $|+i\rangle$ is an eigenstate of the Pauli-Y operator with eigenvalue $+1$).
  4. $T|+i\rangle = \frac{1}{\sqrt{2}}|0\rangle + \left(\frac{-1+i}{2}\right)|1\rangle$.
  5. $Z\left(\frac{1}{\sqrt{2}}|0\rangle + \frac{-1+i}{2}|1\rangle\right) = \frac{1}{\sqrt{2}}|0\rangle + \left(\frac{1-i}{2}\right)|1\rangle = \begin{pmatrix} 0.70710678 \\ 0.5 - 0.5i \end{pmatrix}$.

- **Statevector Amplitudes**:
  - $|0\rangle$ component: $\alpha = \frac{1}{\sqrt{2}} \approx 0.70710678$
  - $|1\rangle$ component: $\beta = 0.5 - 0.5i = \frac{1 - i}{2}$
  - Normalization check: $|\alpha|^2 + |\beta|^2 = \left(\frac{1}{\sqrt{2}}\right)^2 + |0.5 - 0.5i|^2 = 0.5 + 0.5 = 1.0$.

## 6. Conclusion
This experiment demonstrates how a sequence of different single-qubit quantum gates ($H 	o S 	o Y 	o T 	o Z$) transforms a quantum state. By inspecting the exact statevector rather than measuring the qubit, both probability amplitudes and relative complex phases are preserved and observed throughout the entire multi-gate quantum evolution.